drivers/dma/dma_npcx_gdma.c

```diff
@@ -0,0 +1,466 @@
+	uintptr_t dst_addr = cfg->head_block->dest_address;
+
+	/* Check channel is valid */
+	if (channel >= dev_cfg->max_channel) {
+		LOG_ERR("out of range DMA channel %d", channel);
+		return -EINVAL;
+	}
+
+	channel_data = &dev_data->channels[channel];
+
+	k_mutex_lock(&channel_data->ch_mutex, K_FOREVER);
+
+	if (channel_data->channel_active == 1) {
+		LOG_ERR("Channel %x is busy", channel);
+		ret = -EBUSY;
+		goto dma_config_exit;
+	}
+
+	if (cfg->source_burst_length != cfg->dest_burst_length) {
+		LOG_ERR("Burst length mismatch between source and destination");
+		ret = -EINVAL;
+		goto dma_config_exit;
+	}
+
+	/* Check the source address is aligned */
+	if ((src_addr % cfg->source_burst_length) != 0) {
+		LOG_ERR("Source Address Not Aligned (0x%lx)", src_addr);
+		ret = -EINVAL;
+		goto dma_config_exit;
+	}
+
+	/* Check the destination address is aligned */
+	if ((dst_addr % cfg->dest_burst_length) != 0) {
+		LOG_ERR("Destination Address Not Aligned (0x%lx)", dst_addr);
+		ret = -EINVAL;
+		goto dma_config_exit;
+	}
+
+	/* Check the size is aligned */
+	if ((block_size % (cfg->source_burst_length)) != 0) {
+		LOG_ERR("Size Not Aligned");
+		ret = -EINVAL;
+		goto dma_config_exit;
+	}
+
+	/* Check source and destination region not overlay */
+	if (((src_addr + block_size) > dst_addr) && ((dst_addr + block_size) > src_addr)) {
+		LOG_ERR("Transaction Region Overlap");
+		ret = -EFAULT;
+		goto dma_config_exit;
+	}
+
+	channel_data->channel_dir = cfg->channel_direction;
+	channel_data->callback = cfg->dma_callback;
+	channel_data->user_data = cfg->user_data;
+
+	ret = dma_set_controller(dev, channel, cfg);
+
+dma_config_exit:
+	k_mutex_unlock(&channel_data->ch_mutex);
+
+	return ret;
+}
+
+static int dma_npcx_start(const struct device *dev, uint32_t channel)
+{
+	const struct dma_npcx_config *const dev_cfg = dev->config;
+	struct dma_npcx_dev_data *const dev_data = dev->data;
+	struct dma_npcx_ch_data *channel_data;
+	int ret;
+
+	if (channel >= dev_cfg->max_channel) {
+		LOG_ERR("out of range DMA channel %d", channel);
+		return -EINVAL;
+	}
+
+	channel_data = &dev_data->channels[channel];
+
+	k_mutex_lock(&channel_data->ch_mutex, K_FOREVER);
+
+	if (channel_data->channel_active == 1) {
+		LOG_ERR("Channel %x is busy", channel);
+		ret = -EBUSY;
+		goto dma_start_exit;
+	}
+
+	ret = dma_trans_start(dev, channel);
+
+dma_start_exit:
+	k_mutex_unlock(&channel_data->ch_mutex);
+
+	return ret;
+}
+
+static int dma_npcx_stop(const struct device *dev, uint32_t channel)
+{
+	const struct dma_npcx_config *const dev_cfg = dev->config;
+	struct dma_npcx_dev_data *const dev_data = dev->data;
+	struct dma_npcx_ch_data *channel_data;
+
+	if (channel >= dev_cfg->max_channel) {
+		LOG_ERR("out of range DMA channel %d", channel);
+		return -EINVAL;
+	}
+
+	channel_data = &dev_data->channels[channel];
+
+	k_mutex_lock(&channel_data->ch_mutex, K_FOREVER);
+
+	dma_trans_stop(dev, channel);
+
+	channel_data->callback = NULL;
```

Comment:
Calling start on the same transfer after it is completed is allowable, likely a mistake in dropping the callback member here
https://docs.zephyrproject.org/latest/hardware/peripherals/dma.html#driver-implementation-expectations